File: backend/src/common/utils/location_display.py
from __future__ import annotations


def format_location_display(
    province: str | None = None,
    city: str | None = None,
    location: str | None = None,
) -> str:
    province_name = (province or "").strip()
    city_name = (city or "").strip()
    raw_location = (location or "").strip()

    geo_labels = {name for name in (province_name, city_name) if name}

    detail_segments: list[str] = []
    if raw_location:
        normalized = raw_location.replace("،", " — ")
        for segment in normalized.split(" — "):
            cleaned = segment.strip()
            if cleaned and cleaned not in geo_labels and cleaned not in detail_segments:
                detail_segments.append(cleaned)

    parts: list[str] = []
    if province_name:
        parts.append(province_name)
    if city_name and city_name != province_name:
        parts.append(city_name)
    parts.extend(detail_segments)

    return " — ".join(parts)

File: frontend/src/shared/utils/locationDisplay.ts
export function formatLocationDisplay(
  province?: string | null,
  city?: string | null,
  location?: string | null,
): string {
  const provinceName = province?.trim() ?? '';
  const cityName = city?.trim() ?? '';
  const rawLocation = location?.trim() ?? '';

  const geoLabels = new Set([provinceName, cityName].filter(Boolean));

  const detailSegments: string[] = [];
  if (rawLocation) {
    const normalized = rawLocation.replace(/،/g, ' — ');
    for (const segment of normalized.split(' — ')) {
      const cleaned = segment.trim();
      if (cleaned && !geoLabels.has(cleaned) && !detailSegments.includes(cleaned)) {
        detailSegments.push(cleaned);
      }
    }
  }

  const parts: string[] = [];
  if (provinceName) parts.push(provinceName);
  if (cityName && cityName !== provinceName) parts.push(cityName);
  parts.push(...detailSegments);

  return parts.join(' — ');
}

export function getDisasterLocationDisplay(disaster: {
  location_display?: string;
  province?: string;
  city?: string;
  location?: string;
}): string {
  if (disaster.location_display?.trim()) return disaster.location_display;
  return formatLocationDisplay(disaster.province, disaster.city, disaster.location) || '—';
}

export function getMissionLocationDisplay(mission: {
  location_display?: string;
  province?: string;
  city?: string;
  location?: string;
}): string {
  if (mission.location_display?.trim()) return mission.location_display;
  return formatLocationDisplay(mission.province, mission.city, mission.location) || '—';
}
